agent/taskresource/cgroup/cgroup.go

```diff
@@ -44,143 +44,230 @@ var (
+func (cgroup *CgroupResource) TerminalStatus() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(CgroupRemoved)
+}
+
+// GetNextKnownStateProgression returns the state that the resource should
+// progress to based on its `KnownState`.
+func (cgroup *CgroupResource) GetNextKnownStateProgression() taskresource.ResourceStatus {
+	return cgroup.GetKnownStatus() + 1
+}
+
+// ApplyTransition calls the function required to move to the specified status
+func (cgroup *CgroupResource) ApplyTransition(nextState taskresource.ResourceStatus) (bool, error) {
+	transitionFunc, ok := cgroup.resourceStatusToTransitionFunction[nextState]
+	if !ok {
+		seelog.Criticalf("Cgroup Resource [%s]: unsupported desired state transition [%s]: %s",
+			cgroup.taskARN, cgroup.GetName(), CgroupStatus(nextState).String())
+		return false, nil
+	}
+	return true, transitionFunc()
+}
+
+// SteadyState returns the transition state of the resource defined as "ready"
+func (cgroup *CgroupResource) SteadyState() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(CgroupCreated)
+}
+
 // SetKnownStatus safely sets the currently known status of the resource
-func (c *CgroupResource) SetKnownStatus(status taskresource.ResourceStatus) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
+func (cgroup *CgroupResource) SetKnownStatus(status taskresource.ResourceStatus) {
+	cgroup.lock.Lock()
+	defer cgroup.lock.Unlock()
+
+	cgroup.knownStatusUnsafe = status
+	cgroup.updateAppliedStatusUnsafe(status)
+}
 
-	c.knownStatusUnsafe = status
+// updateAppliedStatusUnsafe updates the resource transitioning status
+func (cgroup *CgroupResource) updateAppliedStatusUnsafe(knownStatus taskresource.ResourceStatus) {
+	if cgroup.appliedStatus == taskresource.ResourceStatus(CgroupStatusNone) {
+		return
+	}
+
+	// Check if the resource transition has already finished
+	if cgroup.appliedStatus <= knownStatus {
+		cgroup.appliedStatus = taskresource.ResourceStatus(CgroupStatusNone)
+	}
+}
+
+// SetAppliedStatus sets the applied status of resource and returns whether
+// the resource is already in a transition
+func (cgroup *CgroupResource) SetAppliedStatus(status taskresource.ResourceStatus) bool {
+	cgroup.lock.Lock()
+	defer cgroup.lock.Unlock()
+
+	if cgroup.appliedStatus != taskresource.ResourceStatus(CgroupStatusNone) {
+		// return false to indicate the set operation failed
+		return false
+	}
+
+	cgroup.appliedStatus = status
+	return true
 }
 
 // GetKnownStatus safely returns the currently known status of the task
-func (c *CgroupResource) GetKnownStatus() taskresource.ResourceStatus {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
+func (cgroup *CgroupResource) GetKnownStatus() taskresource.ResourceStatus {
+	cgroup.lock.RLock()
+	defer cgroup.lock.RUnlock()
+
+	return cgroup.knownStatusUnsafe
+}
 
-	return c.knownStatusUnsafe
+// StatusString returns the string of the cgroup resource status
+func (cgroup *CgroupResource) StatusString(status taskresource.ResourceStatus) string {
```

Comment:
Why is this method a member of `CgroupResource` struct? Seems like it can be untethered from it.

Reply:
Each resource has its own status type, which only the resource knows of. Since we use resource abstraction in `task_manager` and each status type has its own string implementation, we call `StatusString` to get the string value of the status. Let me know if you have better suggestions.

Reply:
I see. Can't think of anything better right now. Will keep this in the back of the mind and let you know if I can think of something better.

Reply:
ok, what do you think about adding a new struct called `TestResource` and use that for testing? we cannot really use the mock resource for tests, since they test for the states progression.